### DISCUSSION TITLE
Use function instead of class in really the best solution?

### DISCUSSION THREAD

**Discussion by MasDevProject** [ngrx/platform#4509] [General]
### Which @ngrx/* package(s) are relevant/related to the feature request?

signals

### Information

Hi,
me and my team are implementing the new ```signalStore```.
With very basic components the dev experience is good, but when a component has 5+ methods the store object become hard to read.

This is an example:
```
type MyState = {
  settings: UserSettings | null;
};

const initialState: MyState = {
  settings: null,
};

export const myStore = signalStore(
  withState(initialState),
  withComputed((state) => ({
    comutedTest1: computed(() => state.settings()?.userId.length || 0),
    comutedTest2: computed(() => state.settings()?.userId.length || 0),
    comutedTest3: computed(() => state.settings()?.userId.length || 0),
    comutedTest4: computed(() => state.settings()?.userId.length || 0),
    comutedTest5: computed(() => state.settings()?.userId.length || 0),
  })),
  withMethods(
    (
      store,
      firestore = inject(Firestore),
      identitySvc = inject(...),
      usersSvc = inject(...),
      toastSvc = inject(...),
      electronSvc = inject(...),
      statsSvc = inject(...),
      logSvc = inject(...),
    ) => ({
      async init() {
        try {
          const userId = identitySvc.selectIdentity()!.uid;
          const dataCollection = collection(firestore, 'userSettings');
          const dataQuery = query(dataCollection, where('userId', '==', userId));
          let docs = await getDocs(dataQuery);

          if (!docs.size) {
            await createEmptyEntry(userId, firestore);
            docs = await getDocs(dataQuery);
          }

          const data = docs.docs[0].data() as UserSettings;
          patchState(store, (state) => ({ settings: data }));
        } catch (err) {
          console.log(err);
        }
      },
      async saveUserSettings0() {
        // TODO
        // long method here ....
        // long method here ....
        // long method here ....
        // long method here ....
        // long method here ....
        // long method here ....
        // long method here ....
        // long method here ....
        // long method here ....
        // long method here ....
        // long method here ....
        // long method here ....
      },
      async saveUserSettings1() {
        // TODO
        // long method here ....
        // long method here ....
        // long method here ....
        // long method here ....
        // long method here ....
        // long method here ....
        // long method here ....
        // long method here ....
        // long method here ....
        // long method here ....
        // long method here ....
        // long method here ....
      },
      async saveUserSettings2() {
        // TODO
        // long method here ....
        // long method here ....
        // long method here ....
        // long method here ....
        // long method here ....
        // long method here ....
        // long method here ....
        // long method here ....
        // long method here ....
        // long method here ....
        // long method here ....
        // long method here ....
      },
      async saveUserSettings3() {
        // TODO
        // long method here ....
        // long method here ....
        // long method here ....
        // long method here ....
        // long method here ....
        // long method here ....
        // long method here ....
        // long method here ....
        // long method here ....
        // long method here ....
        // long method here ....
        // long method here ....
      },
    })
  )
);
```

Is this code "safe" to read?
What if I need to use utility methods shared between the methods? Should I create a service/static-method every time?

Isn't a best option to use a class instead? Classes can better arrange the methods and can also use inheritance.

In the last year I can see to much hate for classes. Am I the only one?

### I would be willing to submit a PR to fix this issue

- [ ] Yes
- [X] No

**Comment by timdeschryver** [comment]
Hello, we try to keep the issues section for problems.
Therefore, I'm migrating this issue into a discussion.

**Comment by rainerhahnekamp** [comment]
@MasDevProject, you should extract your long methods to multiple services to shorten the code.

About utility methods, is it about exposing them or not?

Classes vs. Functions. If there is a SignalStore class you have to extend from, that class with all its method ends up in the bundle file, even if you don't use all methods of that SignalStore class. With functions as it is now, only the functions which are really used end up in the bundle.

The main power comes from extensibility. If every extension were a class, you would probably have to add them via inheritance, but more likely, mixins. Mixins in TypeScript are rare in my understanding. Inheritance is - I think - out of question.

---

If the code is too long for a file, you can split up the creation of the SignalStore to multiple files as well. 

---

I don't really see a hate for classes. For the SignalStore, I think it was not possible with the features we have now. For the Analog file format, which uses functions instead of classes for components, I think that's a matter of style, or moving more the template aspect into the focus.

**Comment by DaSchTour** [comment]
Exactly the same question stops me from using signalsStore. In an angular environment it feels wrong to put all this stuff into nested functions.

**Reply by markostanimirovic** [reply]
If you prefer defining stores as classes, `signalState` can be used: https://ngrx.io/guide/signals/signal-state#example-2-signalstate-in-a-service

**Reply by rainerhahnekamp** [reply]
> an angular environment it feels wrong to

At least at the moment. I wouldn't be so sure if that statement has a long life time.

Either way, as Marko said, `signalState` in a class works, but you are loosing all your chances of using the extensions.

**Reply by DaSchTour** [reply]
Well as I already stated many times in the discussion about function vs class-based guards and resolvers the function based approach has limitations and also the changing between class-based and functional mindset leads to junior developers doing confusing things because they don't understand both approaches correctly.

Complex functions using injection also may need to pass the injection to other functions which then makes the stuff really ugly and not any better than when having a class with methods.

Having functions passed into functions that are passed into functions makes code really unreadable and reminds me of the good old jQuery callback times.

Thanks for your help, I'll continue my evaluation of which local state management is the best for my projects.

**Comment by david-bulte** [comment]
What I find a bit annoying, and I think it is a result of this, is that I cannot ask my IDE where a method is being used ("Find Usages" in Intellij).
